### PR TITLE
ReadHistory: refactoring with binary search

### PR DESCRIPTION
**Pull request by hius07**
Get rid of indexing and sorting, reduce flushing.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9603)

### CODE REVIEW

File: frontend/readhistory.lua [hunk 1]
```diff
@@ -50,59 +48,48 @@ local function buildEntry(input_time, input_file)
...
 end
 
-local function timeFirstOrdering(l, r)
-    if l.time == r.time then
-        return l.file < r.file
-    else
-        return l.time > r.time
+function ReadHistory:getIndexByTime(item_time)
+    local hist_num = #self.hist
+    if hist_num == 0 then
```
Comment (poire-z) [inline frontend/readhistory.lua:65]: I don't know if it's just me (and my French), but when I see "num", I think about the "numero" (the _number of a thing_, its index). When I see "nb", it's obvious it's the _number of things_, a count, "nombre" in French.
English may have "number" used for both things, so I dunno :)

File: frontend/readhistory.lua [hunk 2]
```diff
@@ -50,59 +48,48 @@ local function buildEntry(input_time, input_file)
...
         end
     end
-    table.sort(self.hist, timeFirstOrdering)
-    self:_indexing(1)
+    return m + d
 end
 
--- Reduces total count in hist list to a reasonable number by removing last
--- several items.
 function ReadHistory:_reduce()
```
Comment (poire-z) [inline frontend/readhistory.lua:86]: (
I know you're not friend with comments :) and this one might not be needed as the function is small and obvious.
But it was there hurting nobody :)
(and also, given the name of the function, _reduce(), it might help not trying to apprehend it as if it meant "reduce" in the contect of functionnal programming - that I don't know much about :) - and so avoid taking the wrong path trying to figure what it does.)
)

File: frontend/readhistory.lua [hunk 3]
```diff
@@ -148,10 +135,12 @@ function ReadHistory:_readLegacyHistory()
...
                 local file = DocSettings:getNameFromHistory(f)
                 if file ~= nil and file ~= "" then
-                    table.insert(
-                        self.hist,
-                        buildEntry(lfs.attributes(joinPath(history_dir, f), "modification"),
-                                   joinPath(path, file)))
+                    local item_path = joinPath(path, file)
+                    local item_time = lfs.attributes(joinPath(history_dir, f), "modification")
+                    local index = self:getIndexByTime(item_time)
+                    if self.hist[index].file ~= realpath(item_path) then
```
Comment (poire-z) [inline frontend/readhistory.lua:141]: Is that new/added test fixing some issue?

File: frontend/readhistory.lua [hunk 4]
```diff
@@ -227,63 +212,64 @@ function ReadHistory:fileSettingsPurged(path)
...
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == path then
-            self:removeItem(self.hist[i])
-            break
-        end
+    local index = self:getIndexByFile(path)
+    if index then
+        self:removeItem(self.hist[index], index)
     end
-    self:ensureLastFile()
```
Comment (poire-z) [inline frontend/readhistory.lua:245]: Sure this one is not needed ?
(I guess it was me who added all these `self:ensureLastFile()`, may be too many just to be safe :)

File: frontend/readhistory.lua [hunk 5]
```diff
@@ -227,63 +212,64 @@ function ReadHistory:fileSettingsPurged(path)
...
-function ReadHistory:removeItem(item, idx)
-    table.remove(self.hist, item.index or idx)
+function ReadHistory:removeItem(item, idx, no_flush)
+    local index = idx or self:getIndexByTime(item.time)
+    table.remove(self.hist, index)
     os.remove(DocSettings:getHistoryPath(item.file))
-    self:_indexing(item.index or idx)
-    self:_flush()
-    self:ensureLastFile()
+    if not no_flush then
```
Comment (poire-z) [inline frontend/readhistory.lua:254]: I remember we had a talk some months ago about too many flush of something when you would act on many files, possibly with your added multi selection.
Does this no_fluxh is about fixing it ? (I see it's use in clearMissing() only.)

**Comment by NiLuJe** [comment]
This doesn't appear to do binary *insertion* though ;o).

**Comment by hius07** [comment]
> This doesn't appear to do binary _insertion_ though

https://github.com/koreader/koreader/blob/858cdac8900154264feb91d4d531eb5e7d0ff560/frontend/readhistory.lua#L140-L143

**Comment by hius07** [comment]
> At least not in addItem?

`addItem` adds at the top (last opened book).

File: frontend/readhistory.lua (hunk 5 above)
Comment (hius07) [reply frontend/readhistory.lua:254]: It is applpied when removing many (deleted) items from the History list only.

In FM it is not applicable, currently we manipulate on each of the selected files one by one.

**Comment by NiLuJe** [comment]
> `addItem` adds at the top (last opened book).

I don't quite remember how this all works, but this doesn't make much sense (performance-wise, at least).

It'd possibly make better sense to be either properly indexed right then and there (i.e., binary insert), or, at worst, to be *appended* to the table (because that won't require *moving* the whole table, inserting at the front is one of the worst things you can do to a table).

File: frontend/readhistory.lua (hunk 3 above)
Comment (hius07) [reply frontend/readhistory.lua:141]: Yes, previously items were just added, then sorted, and then duplicates were removed.
Now we check for existence and add legacy history items if they has not been added earlier.

File: frontend/readhistory.lua (hunk 4 above)
Comment (hius07) [reply frontend/readhistory.lua:245]: Yes, we do that in `removeItem`.

Comment (poire-z) [reply frontend/readhistory.lua:245]: OK.
(So, shouldn't you remove it also from inside ReadHistory:fileDeleted()? Or do it only in the dim = true branch if needed ?)

**Comment by hius07** [comment]
> It'd possibly make better sense to be either properly indexed right then and there (i.e., binary insert), or, at worst, to be _appended_ to the table (because that won't require _moving_ the whole table: inserting at the front is one of the worst things you can do to a table).

Currently the history list is stored as an array, [1] is the latest book.
How to deal with it without migration to some new format?

**Comment by poire-z** [comment]
I also thought it would require a migration.
I'd say: don't bother - inserting at start of the history table is nothing compared to everything else happening on new book opening :)
(But @NiLuJe being in a perfectionnist mayhem, I might no be heard :)

**Comment by NiLuJe** [comment]
I was indeed advocating for migrating to a better format ;o).

(At which point... SQL might seriously start to make mire sense ;p).

**Comment by poire-z** [comment]
(My :-1: is mostly and seriously only about the SQL bit :)

**Comment by NiLuJe** [comment]
I'm not actually sure the internal format matters all that much, though: just start pushing stuff to the tail now and make the code loop in reverse?

But if you're worried about that, we can stick a dumb migration in onetimemigration?

**Comment by poire-z** [comment]
It's just that before appending at the end and doing the iteration in reverse, we need to once revert the whole list, so a boring onetime migration _is_ required.
And there might be external tools/scripts that use our current history.lua format (@noembryo : does KOHighlight read it ?) - although they will find what they want and might not care about the order.

**Comment by hius07** [comment]
About binary: `getIndexByTime(time)` returns index of the item if it exists, or insertion index otherwise.

**Comment by NiLuJe** [comment]
> > @noembryo : does KOHighlight read it ?
> 
> The `history.lua` file is not read by KOHighlights. It only reads any `.lua` files that are in the history folder (if it exists).

That's good to know, thanks ;).

(Means we might be able to cadge some performance out of flush by switching to a binary Persist format... ;p).

**Comment by poire-z** [comment]
> (Means we might be able to cadge some performance out of flush by switching to a binary Persist format... ;p).

Grrrrrr.
Btw, as you're so into dumping : is serpent or any other able to dump arrays as `{ item1, item2 }` instead of the current `{ [1] = item1, [2] = item2 }`. The latter is a pain when you need to edit/remove stuff manually, as you need to renumber all the indexes (which I sometimes want to do when, for example, after I had checked dozens of html files from a testsuite).

Also in there, when testing - but we could easily clean that one :):
https://github.com/koreader/koreader/blob/7e209139d5253d34272adf1f5071a74592a2f4e4/frontend/ui/font.lua#L107-L116

File: frontend/readhistory.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/readhistory.lua]: > I don't know if it's just me (and my French), but when I see "num", I think about the "numero" (the number of a thing, its index).

Yes. :-) To me "numero" is just a funny way of saying number.

**Comment by NiLuJe** [comment]
> is serpent or any other able to dump arrays as `{ item1, item2 }` instead of the current `{ [1] = item1, [2] = item2 }`.

Yeah, in fact I don't think serpent supports the array notation *at all* (it only uses it as necessary for string keys) but I don't think there is currently a setting to force it, it defaults to the more compact syntax.

File: frontend/readhistory.lua [hunk 6]
```diff
@@ -268,20 +268,24 @@ end
...
     if file ~= nil and lfs.attributes(file, "mode") == "file" then
-        local index = self:getIndexByFile(realpath(file))
-        if index then
-            table.remove(self.hist, index)
-        end
-        local now = ts or os.time()
-        table.insert(self.hist, 1, buildEntry(now, file))
         -- util.execute("/bin/touch", "-a", file)
         -- This emulates `touch -a` in LuaFileSystem's API, since it may be absent (Android)
         -- or provided by busybox, which doesn't support the `-a` flag.
```
Comment (NiLuJe) [inline frontend/readhistory.lua:273]: Unrelated to your endeavors, but feel free to remove that comment, because using lfs was always the sanest approach vs. spawning `touch`, regardless of the platform ;).

**Comment by poire-z** [comment]
> Yeah, in fact I don't think serpent supports the array notation at all (it only uses it as necessary for string keys) but I don't think there is currently a setting to force it, it defaults to the more compact syntax.

I still see ` {[1] = ` in my settings.reader.lua and metadata.epub.lua, ie (and others):
```lua
    ["word_spacing"] = {
        [2] = 90,
        [1] = 100,
    },
```
Isn't your usage of serpent already merged ?

**Comment by poire-z** [comment]
There is one readhistory test that has always occasionally bugged us with random failures because of something like you mentionned. I don't exactly remember which, possibly mentionned among https://github.com/koreader/koreader/issues?q=sort%3Aupdated-desc+readhistory+unit+test ...

Anyway, if it was just me, I would exterminate all unit tests :) so you can start with the ones that bug you :)

**Comment by NiLuJe** [comment]
> > Yeah, in fact I don't think serpent supports the array notation at all (it only uses it as necessary for string keys) but I don't think there is currently a setting to force it, it defaults to the more compact syntax.
> 
> I still see `{[1] =` in my settings.reader.lua and metadata.epub.lua, ie (and others):
> 
> ```lua
>     ["word_spacing"] = {
>         [2] = 90,
>         [1] = 100,
>     },
> ```
> 
> Isn't your usage of serpent already merged ?

I left LuaSettings & co alone for now ;o).

**Comment by NiLuJe** [comment]
> Regarding unit tests: they require items with the same time to be sorted alphabetically based on filename. How can we get two history items with the same time? Maybe, after the system clock failure? Should we consider this case?

```
┌─(niluje@illyria:pts/11)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/var/tmp/niluje)─┐
└─(0.85:36%:04:43:%)── touch foo bar baz                                                                                                                                                                                                                                                                                                               ──(Fri, Oct 07)─┘
┌─(niluje@illyria:pts/11)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/var/tmp/niluje)─┐
└─(0.20:37%:17:19:%)── stat foo bar baz                                                                                                                                                                                                                                                                                                                ──(Fri, Oct 07)─┘
  File: foo
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 0,34    Inode: 1150969     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  niluje)   Gid: ( 1000/  niluje)
Access: 2022-10-07 17:19:01.990498163 +0200
Modify: 2022-10-07 17:19:01.990498163 +0200
Change: 2022-10-07 17:19:01.990498163 +0200
 Birth: 2022-10-07 17:19:01.990498163 +0200
  File: bar
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 0,34    Inode: 1150970     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  niluje)   Gid: ( 1000/  niluje)
Access: 2022-10-07 17:19:01.990498163 +0200
Modify: 2022-10-07 17:19:01.990498163 +0200
Change: 2022-10-07 17:19:01.990498163 +0200
 Birth: 2022-10-07 17:19:01.990498163 +0200
  File: baz
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 0,34    Inode: 1150971     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/  niluje)   Gid: ( 1000/  niluje)
Access: 2022-10-07 17:19:01.990498163 +0200
Modify: 2022-10-07 17:19:01.990498163 +0200
Change: 2022-10-07 17:19:01.990498163 +0200
 Birth: 2022-10-07 17:19:01.990498163 +0200
```

It's not entirely out of the realm of possibilities, especially for people using sync'ing methods (and, also, FAT32 has terrible time resolution, so it miiiight happen in less "synthetic" workflows ;p).

**Comment by NiLuJe** [comment]
> There is one readhistory test that has always occasionally bugged us with random failures because of something like you mentionned. I don't exactly remember which, possibly mentionned among https://github.com/koreader/koreader/issues?q=sort%3Aupdated-desc+readhistory+unit+test ...
> 
> Anyway, if it was just me, I would exterminate all unit tests :) so you can start with the ones that bug you :)

It used to, but we fixed that when MockTime was introduced, IIRC ;).

**Comment by NiLuJe** [comment]
TL;DR: Regardless of unit tests (of which many indeed do weird and nonsensical things), I think it *does* make sense to have a deterministic behavior for files with identical timestamps (whatever the method actually is, and alpha sort doesn't sound too bad; one other contender could be insertion order?).

**Comment by NiLuJe** [comment]
> I left LuaSettings & co alone for now ;o).

The (main) reasoning was that it would cause a massive diff, which might be annoying for people with custom workflows.

File: frontend/readhistory.lua [hunk 7]
```diff
@@ -138,9 +138,15 @@ function ReadHistory:_readLegacyHistory()
                 local file = DocSettings:getNameFromHistory(f)
                 if file ~= nil and file ~= "" then
                     local item_path = joinPath(path, file)
+                    local real_path = realpath(item_path)
                     local item_time = lfs.attributes(joinPath(history_dir, f), "modification")
                     local index = self:getIndexByTime(item_time)
-                    if self.hist[index].file ~= realpath(item_path) then
+                    if self.hist[index].file ~= real_path then
+                        while index <= #self.hist -- sort alhabetically
```
Comment (zwim) [inline frontend/readhistory.lua:145]: Nit: alphabetically

File: frontend/readhistory.lua [hunk 8]
```diff
@@ -70,7 +70,7 @@ function ReadHistory:getIndexByTime(item_time)
     elseif item_time < self.hist[hist_nb].time then
         return hist_nb + 1
     end
-    -- binary search
+    -- binary search (returns min index when several elements equal item_time)
     local s, e, m, d = 1, hist_nb
```
Comment (zwim) [inline frontend/readhistory.lua:74]: Line 76 can be optimized like
m = bit.rshift(e + s, 1)

And the elseif can be turned into an else in line 79.

Comment (hius07) [reply frontend/readhistory.lua:74]: Thanks

**Comment by hius07** [comment]
Since 2016 (https://github.com/koreader/koreader/pull/1977) nothing is written to the /history folder.
On every start `readhistory` still adds legacy history items to the history list and then removes duplicates.
Can we stop reading legacy history folder?

**Comment by NiLuJe** [comment]
Isn't it still populated for books read on a RO media? (I keep forgetting how this works :s).

**Comment by poire-z** [comment]
> On every start readhistory still adds legacy history items to the history list and then removes duplicates.

Aren't the legacy items cleaned up once added (and later flushed to our current history.lua) ?
If not, I guess you could clean them.

> Can we stop reading legacy history folder?

It's either:
- keep reading it, but clean it (as mentionned above), so next time, we'll read it and will find nothing
- a one-time migration

Because if some old users upgrade from KOReader 2014, we don't want to lose their history - we're that friendly ! :)

**Comment by poire-z** [comment]
> Isn't it still populated for books read on a RO media? (I keep forgetting how this works :s).

Me too :/ We indeed store the metadata.epub.lua in history/that.book.epub.lua if we can't create that.book.sdr/metadata.epub.lua.
I don't know if it is the presence of history/that.book.epub.lua that makes that.book.epub a legacy history item - or if it's something else.

**Comment by NiLuJe** [comment]
> > Isn't it still populated for books read on a RO media? (I keep forgetting how this works :s).
> 
> Me too :/ We indeed store the metadata.epub.lua in history/that.book.epub.lua if we can't create that.book.sdr/metadata.epub.lua. I don't know if it is the presence of history/that.book.epub.lua that makes that.book.epub a legacy history item - or if it's something else.

As far as DocSettings is concerned, it's purely based on whether it can write the "proper" one or not:

https://github.com/koreader/koreader/blob/93ae341823c4e1f796fab4abcbf0129de77f093a/frontend/docsettings.lua

**Comment by NiLuJe** [comment]
And I *think* the priority sorting is broken, because table.sort stops on the first nil, and the code happily inserts nils in the table... ;o].

File: frontend/readhistory.lua [hunk 9]
```diff
@@ -178,129 +179,131 @@ function ReadHistory:getLastFile()
...
-    self:ensureLastFile()
 end
 
-function ReadHistory:updateItemByPath(old_path, new_path)
-    for i = #self.hist, 1, -1 do
-        if self.hist[i].file == old_path then
-            self.hist[i].file = new_path
-            self.hist[i].text = new_path:gsub(".*/", "")
-            self:_flush()
-            self:reload(true)
```
Comment (poire-z) [inline frontend/readhistory.lua:254]: Any reason for you to have moved this function up in the file (which make it harder to grasp the changes) ? (Asking in case this was not voluntarily.)
(Unless it is just github/git diff that has a hard time finding what has changed.)

Comment (hius07) [reply frontend/readhistory.lua:254]: To place funtions more logically: `removeItem` near `removeItemByPath`.
(I'm still struggling with spec, please do not review intermediate commits)

**Comment by zwim** [comment]
@hius07 Do you know, that you can run the tests on you computer? Sou you could maybe save some uploadings 

Just in case you have missed this development feature:
```
./kodev activate   % do this once per session
./kodev test front % run all tests (takes some time)
./kodev test front readerhistory_spec.lua % run just this test (fast)
```

**Comment by hius07** [comment]
> you can run the tests on you computer?

Thank you, I do not have local repository.

**Comment by hius07** [comment]
I believe it is good now.

**Comment by NiLuJe** [comment]
While I'm far from being the biggest fan of the testsuite, I vaguely recall having done a pass on this one, and it somewhat making sense, so... I wouldn't mind an explanation for all the torn out stuff ;).

**Comment by hius07** [comment]
> an explanation for all the torn out stuff

1) `addItem`  had a second argument that was used in the unit tests only. I had removed it so changed the tests respectively. If it is a good practice to have such a code, I will restore.
2) A couple of tests was almost duplicated.
3) Pieces of code about indexing and sorting have been removed, no need for tests.

**Review by poire-z — APPROVED** [review]
Nothing looks odd - but I haven't really focused on all the bits, so trusting you.
May be for v2022.11 - if v2022.10 is near ?

**Comment by hius07** [comment]
Yes, 2022.11, let's live with it for a while.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r13, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9603)* status: all files reviewed, 9 unresolved discussions (waiting on @hius07, @poire-z, and @zwim)

___
*[`frontend/readhistory.lua` line 84 at r15](https://reviewable.io/reviews/koreader/koreader/9603#-NDurDzaDO5YB9kZEs_4:-NDurDzaDO5YB9kZEs_5:b-o9p8l9) ([raw file](https://github.com/koreader/koreader/blob/a93441d39ddb07c1c4e7548ad5f1bb557b0feecf/frontend/readhistory.lua#L84)):*
> ```Lua
>         else
>             e, d = m - 1, 0
>         end
> ```

Shouldn't this be the other way around?

(i.e., `if item_time < mid_time then e = m - 1`) (or `item_time > mid_time`, but since Lua prefers lt for comparisons, I kinda prefer keeping to the < approach ;)))



_Code quote:_
```Lua
        if item_time < self.hist[m].time then
            s, d = m + 1, 1
        else
            e, d = m - 1, 0
        end
```
___
*[`frontend/readhistory.lua` line 91 at r15](https://reviewable.io/reviews/koreader/koreader/9603#-NDus8oGDjg-agYmuCXu:-NDus8oGDjg-agYmuCXv:b6smfp5) ([raw file](https://github.com/koreader/koreader/blob/a93441d39ddb07c1c4e7548ad5f1bb557b0feecf/frontend/readhistory.lua#L91)):*
> ```Lua
> --- Reduces number of history items to the required limit by removing old items.
> function ReadHistory:_reduce()
>     local history_size = G_reader_settings:readSetting("history_size") or 500
> ```

I'd possibly add a comment that this setting has no UI ;).

**Comment by NiLuJe** [comment]
> 1. `addItem`  had a second argument that was used in the unit tests only. I had removed it so changed the tests respectively. If it is a good practice to have such a code, I will restore.

Okay, now I remember ;o).

That's from https://github.com/koreader/koreader/pull/6991, and it was the only way to make testing most mtime stuff (e.g., collisions) feasible ;).

It's a bit (a lot ;p) clunky, granted, and, as we've discussed earlier, mtime collisions are hard to actually trip in practice, but not entirely out of the realm of possibilities, so, dunno, I can't quite recall how critical this all really is ;).

**Comment by hius07** [comment]
> Shouldn't this be the other way around?
> 
> (i.e., `if item_time < mid_time then e = m - 1`)

History is sorted in reverse order, hist[m-1].time > hist[m].time

**Comment by hius07** [comment]
> That's from #6991, and it was the only way to make testing most mtime stuff (e.g., collisions) feasible ;).

With the new optimized `addItem` inserting the item at certain time position requires a separate chunk of code. 
Should I add it for test purposes?

**Comment by NiLuJe** [comment]
> > Shouldn't this be the other way around?
> > (i.e., `if item_time < mid_time then e = m - 1`)
> 
> History is sorted in reverse order, hist[m-1].time > hist[m].time

Yeah, I just had the same epiphany, thanks ;).

**Comment by NiLuJe** [comment]
> With the new optimized `addItem` inserting the item at certain time position requires a separate chunk of code.
> Should I add it for test purposes?

On the one hand, it's boring and clunky; on the other... you never know what might happen ;o).

No strong feelings, but if it doesn't mangle the code too much, why not?

**Comment by hius07** [comment]
All original unit tests restored and passed.
